### PR TITLE
Note additions

### PR DESCRIPTION
**Pull request by Schottkyc137**
see issue: #221, #381, #145 

### Visible/Frontend Changes

- [x] Meeting Notes now have a table of children that were present/absent
- [x] Users can choose whether a child was present or absent and set remarks 
- [x] This table is suitable for small screens as well as big screens
- [x] The absence/presence of a single child will be reflected by suitable colors in the child's `notes-list` component
- [x] The `Note-Manager`-component uses a paginator

### Architectural/Backend Changes
- [x] `Note` now is now it's own module and is no longer a sub-component of the `ChildrenModule`
- [x] A `NotesService` has been implemented to fetch resources from the `EntityMapper` and distribute updates throughout every listening component
- [x] Minor changes have been made to the `EntitySubrecord` and `ChildSelect`-components to enable these new features
- [x] Demo-Data now also incorporates a generator for meeting-notes where certain children are marked as absent/present with suitable remarks

### CODE REVIEW

**Comment by sleidig** [comment]
sorry for the merge conflicts with the master branch, this will be mostly due to the moving of files into a new folder structure (#394).

Is this otherwise complete and ready for review and merge?

**Comment by Schottkyc137** [comment]
Concerning features and other changes, this is done.
Currently, I'm only working on improving the test coverage & fixing other codeclimate errors

File: src/app/child-dev-project/children/children.service.ts [hunk 1]
```diff
@@ -4,7 +4,7 @@ import { Child } from './model/child';
 import { EntityMapperService } from '../../core/entity/entity-mapper.service';
 import { AttendanceMonth } from '../attendance/model/attendance-month';
 import { Database } from '../../core/database/database';
-import { Note } from '../notes/model/note';
+import { NoteModel } from '../notes/note.model';
```
Comment (sleidig) [inline src/app/child-dev-project/children/children.service.ts]: Any particular reason to rename the `Note` class to `NoteModel`? I think we should keep this consistent for all Entity classes and personally favour the short `Note` but we can put this up for discussion

File: src/app/child-dev-project/children/notes-list/notes-list.component.ts [hunk 2]
```diff
@@ -30,36 +33,35 @@ export class NotesComponent implements OnInit {
...
 
   ngOnInit() {
     this.route.paramMap.subscribe(params => {
       this.childId = params.get('id');
+    });
 
-      this.childrenService.getNotesOfChild(this.childId)
-        .subscribe(results => this.records = results.sort((a, b) => {
-          return (b.date ? b.date.valueOf() : 0) - (a.date ? a.date.valueOf() : 0); } ));
+    this.entityMapperService.loadType<NoteModel>(NoteModel).then(notes => {
```
Comment (sleidig) [inline src/app/child-dev-project/children/notes-list/notes-list.component.ts]: Loading every note and doing in-memory filtering may become a performance issue. `ChildrenService.getNotesOfChild()` uses a pouchdb database index and should (in theory) be the better choice.

Did you have any particular issues that pushed you to change this here?

File: src/app/child-dev-project/notes/attendance.model.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
+
+export class AttendanceModel {
```
Comment (sleidig) [inline src/app/child-dev-project/notes/attendance.model.ts]: This sounds very similar to our `AttendanceMonth` class and the (completely separate) system for keeping students' school attendance. Maybe `MeetingNoteAttendance` or something like that can make its purpose more clear?

File: src/app/child-dev-project/notes/note-detail/note-detail.component.html [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+                               fxFlex>
+      </app-child-presence-list>
+      <app-child-presence-list [note]="entity"
+                               [recordForm]="recordForm"
+                               [present]="false"
+                               [label]="'Absent'"
+                               class="padding-bottom"
+                               fxFlex>
+      </app-child-presence-list>
+    </div>
```
Comment (sleidig) [inline src/app/child-dev-project/notes/note-detail/note-detail.component.html]: Let's extract this into a separate component as well so the responsive version can be easily re-used (and the template here will be shorter and more manageable). Possibly such a new component could also include the `app-child-select` component above and be somewhat of an alternative to the simple `app-child-select`

File: src/app/core/ui-helper/AbstractDetailsComponent.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
+import { Inject, Input, OnInit, ViewChild } from '@angular/core';
+import { MAT_DIALOG_DATA, MatDialogRef } from '@angular/material/dialog';
+import { ConfirmationDialogService } from './confirmation-dialog/confirmation-dialog.service';
+import { EntityMapperService } from '../entity/entity-mapper.service';
+import { Entity } from '../entity/entity';
+
+export abstract class AbstractDetailsComponent<T extends Entity> implements OnInit {
```
Comment (sleidig) [inline src/app/core/ui-helper/AbstractDetailsComponent.ts]: I like the idea to factor out that abstract class :+1:

File: src/app/child-dev-project/children/children.service.ts (hunk 1 above)
Comment (Schottkyc137) [reply src/app/child-dev-project/children/children.service.ts]: I have done this to distinguish the new `NoteModel` from the old `Note` during refactoring from a component to a module. I kept the change to make it clear that this was a model. Since I agree that this is no longer needed, I will revert the changes

File: src/app/child-dev-project/children/notes-list/notes-list.component.ts (hunk 2 above)
Comment (Schottkyc137) [reply src/app/child-dev-project/children/notes-list/notes-list.component.ts]: I had an issue with this but could resolve this in another way. Until now I thought that these two ways of getting the data were equal in performance. If that's not the case, I will revert these changes as well

Comment (Schottkyc137) [reply src/app/child-dev-project/children/notes-list/notes-list.component.ts]: Looking at the code right now, the issue was not resolved. `ChildrenService` filters if the children-array of the Note contains the child-id (which is never true since the Array is of type `AttendanceModel`). Is it possible to change the database quarry to check if any of it's `AttendanceModel`s contains the id?

Comment (sleidig) [reply src/app/child-dev-project/children/notes-list/notes-list.component.ts]: For the PouchDb/CouchDB based Database queries should be more efficient, the database then only loads the required documents and returns them (and hopefully the developers of pouchdb have figured out a way to do that reasonably efficient).

The query index is also created in the ChildrenService:
https://github.com/Aam-Digital/ndb-core/blob/134d808a86fde5983c81e93f2ac1e6d0a6bba98b/src/app/child-dev-project/children/children.service.ts#L229-L243

Beware, unfortunately the function has to be a string containing the javascript function (to prevent the build optimization from messing this up). You can also refer to the PouchDB guide about how those query functions work: https://pouchdb.com/guides/queries.html

Basically you'll just have to change that function to `emit()` the childId from its new nested place.

**Comment by sleidig** [comment]
closing this as PR #431 is built upon this branch

**Comment by sleidig** [comment]
closing this as PR #431 is built upon this branch
